### ISSUE TITLE
ImportError: Unsafe global: copy_reg._reconstructor

### ISSUE THREAD

**Issue by papachoco** [zopefoundation/ZEO#113]
We are trying ZEO 5.1.2 and we got the error below. We are using ZODB 5.3.0

Carlos
```
2018-03-27 13:25:17,315 ERROR [ZEO.asyncio.marshal][4804095152:2209][Users zeo client networking thread] can't decode message: '((ccopy_reg\n_reconstructor\n(czodbpickle\nbinary\nc__b...'
2018-03-27 13:25:17,317 ERROR [ZEO.asyncio.base][4804095152:2209][Users zeo client networking thread] data_received 4 0 True
Traceback (most recent call last):
  File "/Users/Birdwell/Projects/nti.dataserver-buildout/eggs/ZEO-5.1.2-py2.7.egg/ZEO/asyncio/base.py", line 128, in data_received
    self.message_received(collected)
  File "/Users/Birdwell/Projects/nti.dataserver-buildout/eggs/ZEO-5.1.2-py2.7.egg/ZEO/asyncio/client.py", line 206, in message_received
    msgid, async, name, args = self.decode(data)
  File "/Users/Birdwell/Projects/nti.dataserver-buildout/eggs/ZEO-5.1.2-py2.7.egg/ZEO/asyncio/marshal.py", line 91, in pickle_decode
    return unpickler.load() # msgid, flags, name, args
  File "/Users/Birdwell/Projects/nti.dataserver-buildout/eggs/ZEO-5.1.2-py2.7.egg/ZEO/asyncio/marshal.py", line 155, in find_global
    raise ImportError("Unsafe global: %s.%s" % (module, name))
ImportError: Unsafe global: copy_reg._reconstructor
```

**Comment by jamadden** [comment]
I suspect you're using an existing Data.fs that was used with ZODB 5.4?

I have a fix for this in #110 (although it could be pulled out separately if needed and desired).

@jimfulton There is at least [one place](https://github.com/zopefoundation/ZEO/blob/master/src/ZEO/asyncio/marshal.py#L58) where ZEO is choosing protocol 1 on Python 2 for sending messages.  Do you think we should bump that to 3 like ZODB?

**Comment by jimfulton** [comment]
On Wed, Mar 28, 2018 at 7:33 AM, Jason Madden <notifications@github.com>
wrote:

> I suspect you're using an existing Data.fs that was used with ZODB 5.4?
>
Why would that matter?

> I have a fix for this in #110
> <https://github.com/zopefoundation/ZEO/pull/110> (although it could be
> pulled out separately if needed and desired).
>
What is the issue?

> @jimfulton <https://github.com/jimfulton> There is at least one place
> <https://github.com/zopefoundation/ZEO/blob/master/src/ZEO/asyncio/marshal.py#L58>
> where ZEO is choosing protocol 1 on Python 2 for sending messages. Do you
> think we should bump that to 3 like ZODB?
>
IDK.  I was wondering why zodbpickle.binary was an issue. I guess some
object ids are being set to that.

I suppose that protocol 3 would be useful in the long term, but I worry
that we're flailing.

I'm sorely tempted to revert the ZODB change until we understand the impact
better. I'm also at a loss because I have very little time during the week
to focus on this.

**Comment by jamadden** [comment]
>> I suspect you're using an existing Data.fs that was used with ZODB 5.4?
>Why would that matter?

Because that's most likely how `zodbpickle.binary` got in there.

> What is the issue?

The issue is `zodbpickle.binary` pickled in protocol 1 on Python 2; that's how references to `copy_reg._reconstructor` get in. Before the fix in #110, `copy_reg` wasn't allowed to be used.

> I suppose that protocol 3 would be useful in the long term, but I worry that we're flailing.

I think that it's probably useful in the short term too.

> I'm sorely tempted to revert the ZODB change until we understand the impact better.

That won't help people who already have OIDs in their database as instances of `zodbpickle.binary`. The ZEO changes will still be necessary. I'm afraid the only way out is through 🤷‍♂️ 

Personally, I'd like to see #110 merged, and then a quick follow-up PR that uses protocol 3 merged, and then a release. But that's just me 😄

**Comment by jimfulton** [comment]
On Wed, Mar 28, 2018 at 7:49 AM, Jason Madden <notifications@github.com>
wrote:

> I suspect you're using an existing Data.fs that was used with ZODB 5.4?
> Why would that matter?
>
> Because that's most likely how zodbpickle.binary got in there.
>
> What is the issue?
>
> The issue is zodbpickle.binary pickled in protocol 1 on Python 2; that's
> how references to copy_reg._reconstructor get in. Before the fix in #110
> <https://github.com/zopefoundation/ZEO/pull/110>, copy_reg wasn't allowed
> to be used.
>
> I suppose that protocol 3 would be useful in the long term, but I worry
> that we're flailing.
>
> I think that it's probably useful in the short term too.
>
> I'm sorely tempted to revert the ZODB change until we understand the
> impact better.
>
> That won't help people who already have OIDs in their database as
> instances of zodbpickle.binary. The ZEO changes will still be necessary.
> I'm afraid the only way out is through 🤷‍♂️
>
Personally, I'd like to see #110
> <https://github.com/zopefoundation/ZEO/pull/110> merged,
>
Done.

> and then a quick follow-up PR that uses protocol 3 merged, and then a
> release. But that's just me 😄
>
I defer, because I don't have time to do anything else until this weekend.
:/

OK. Let's see of we can get all green on tests before release by restarting
failed travis runs, where necessary.

Thanks!!!

**Comment by jamadden** [comment]
Thanks!

#114 is the wire protocol change.
